src/main/java/org/apache/mesos/scheduler/DefaultScheduler.java

```diff
@@ -0,0 +1,309 @@
+package org.apache.mesos.scheduler;
+
+import com.google.protobuf.TextFormat;
+import org.apache.mesos.Protos;
+import org.apache.mesos.Scheduler;
+import org.apache.mesos.SchedulerDriver;
+import org.apache.mesos.curator.CuratorStateStore;
+import org.apache.mesos.dcos.DcosConstants;
+import org.apache.mesos.offer.*;
+import org.apache.mesos.reconciliation.DefaultReconciler;
+import org.apache.mesos.reconciliation.Reconciler;
+import org.apache.mesos.scheduler.plan.*;
+import org.apache.mesos.scheduler.recovery.*;
+import org.apache.mesos.scheduler.recovery.constrain.LaunchConstrainer;
+import org.apache.mesos.scheduler.recovery.constrain.TimedLaunchConstrainer;
+import org.apache.mesos.scheduler.recovery.monitor.TimedFailureMonitor;
+import org.apache.mesos.specification.DefaultPlanSpecificationFactory;
+import org.apache.mesos.specification.PlanSpecification;
+import org.apache.mesos.specification.ServiceSpecification;
+import org.apache.mesos.state.PersistentOperationRecorder;
+import org.apache.mesos.state.StateStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This scheduler when provided with a ServiceSpecification will deploy the service and recover from encountered faults
+ * when possible.  Changes to the ServiceSpecification will result in rolling configuration updates, or the creation of
+ * new Tasks where applicable.
+ */
+public class DefaultScheduler implements Scheduler {
+    private static final Integer DELAY_BETWEEN_DESTRUCTIVE_RECOVERIES_SEC = 10 * 60;
+    private static final Integer PERMANENT_FAILURE_DELAY_SEC = 20 * 60;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final ServiceSpecification serviceSpecification;
+    private final String zkConnectionString;
+    private final ExecutorService executor = Executors.newFixedThreadPool(1);
```

Comment:
I was more concerned about handling status updates vs handling offers in the same thread.

I think handling offer batches one at a time is better and keeps things simple.